### PR TITLE
Implementing autocompletion for JetBrains lookup manager

### LINKED ISSUE #57335: JetBrains: integrate autocomplete with lookup manager
Currently, autocomplete in VS Code has a tight integration with the editor's "suggest widget" (aka. lookup manager in IntelliJ). We should update autocomplete in IntelliJ to mirror the VS Code behavior and, optionally, make it even better than in VS Code 😏 We have more control over autocomplete in IntelliJ so maybe we can do an even tighter integration with the lookup manager. We can optionally consider adding a setting to enable/disable the integration with the lookup manager.

Example of Cody autocomplete working with VS Code lookup manager

Image

### PR DESCRIPTION
**Pull request by Sa1to**
It fixes [#57335](https://github.com/sourcegraph/sourcegraph/issues/57335)
## Test plan
* Having enabled autocompletion, trigger lookup manager (for example: by pressing `control+space`) to get hint for possible methods:

https://github.com/sourcegraph/jetbrains/assets/9321940/0e9d9bb9-cc28-4669-a5ee-1711425cbbbd

* Having enabled autocompletion, trigger lookup manager (for example: by pressing `control+space`) to get hint for possible classes:

https://github.com/sourcegraph/jetbrains/assets/9321940/b154414c-906d-4e09-97c7-f7f801fd0fd8

* Disable autocomplete for lookup elements and try to trigger lookup autocompletion again

https://github.com/sourcegraph/jetbrains/assets/9321940/b9eb676b-41bb-494b-9758-769be208b998

### CODE REVIEW

File: client/jetbrains/src/test/kotlin/com/sourcegraph/cody/chat/CodyChatMessageHistoryTest.kt [hunk 1]
```diff
@@ -30,7 +24,6 @@ class CodyChatMessageHistoryTest {
     history.messageSent(textArea)
     textArea.text = "test 2"
     history.messageSent(textArea)
-
     history.popUpperMessage(textArea)
```
Comment (exigow) [inline client/jetbrains/src/test/kotlin/com/sourcegraph/cody/chat/CodyChatMessageHistoryTest.kt:26]: Is this related? It looks like stray code.

File: src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt [hunk 2]
```diff
@@ -275,15 +278,43 @@ class CodyAutocompleteManager {
...
+      val lastCommonSuffixCharacterPosition =
+          findLastCommonSuffixElementPosition(
+              originalText.substring(0, caretPositionInLine), lookupString)
+      insertTextFirstLine =
+          insertTextFirstLine.removeRange(lastCommonSuffixCharacterPosition, caretPositionInLine)
+
+      // For each autocomplete item, remove common part with already inserted part of the lookup
+      // element.
+      items.stream().forEach { item ->
+        run {
```
Comment (exigow) [inline src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt]: Is it necessary to place this code in run? It seems to me that it doesn't make a difference.

File: src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt [hunk 3]
```diff
@@ -124,7 +119,12 @@ class CodyAutocompleteManager {
    * @param editor The editor instance to provide autocomplete for.
    * @param offset The character offset in the editor to trigger auto-complete at.
    */
-  fun triggerAutocomplete(editor: Editor, offset: Int, triggerKind: InlineCompletionTriggerKind) {
+  fun triggerAutocomplete(
+      editor: Editor,
+      offset: Int,
+      triggerKind: InlineCompletionTriggerKind,
+      lookupString: String?
```
Comment (exigow) [inline src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt]: After taking a look at the usages, I have a small suggestion: `lookupString: String? = null`.

File: src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt [hunk 4]
```diff
@@ -143,7 +143,8 @@ class CodyAutocompleteManager {
       return
     }
     val currentCommand = CommandProcessor.getInstance().currentCommandName
-    if (isTriggeredImplicitly && isCommandExcluded(currentCommand)) {
+    if (isTriggeredImplicitly &&
+        (lookupString.isNullOrEmpty() && isCommandExcluded(currentCommand))) {
```
Comment (exigow) [inline src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt]: Parentheses are not necessary in this case.

File: src/main/kotlin/com/sourcegraph/cody/vscode/InlineAutocompleteItem.kt [hunk 5]
```diff
@@ -1,8 +1,12 @@
...
+    var insertText: String,
     val filterText: String,
-    val range: Range,
+    var range: Range,
     val command: Command
-)
+) {
+  override fun toString(): String {
+    return "InlineAutocompleteItem(insertText='$insertText', filterText='$filterText', range=$range, command=$command)"
+  }
```
Comment (exigow) [inline src/main/kotlin/com/sourcegraph/cody/vscode/InlineAutocompleteItem.kt]: Do we need `toString`?

Comment (Strum355) [reply src/main/kotlin/com/sourcegraph/cody/vscode/InlineAutocompleteItem.kt]: If we change this to a data class, we could get it automatically. Do `Range` or `Command` prohibit this being a data class?

File: src/main/kotlin/com/sourcegraph/cody/vscode/InlineAutocompleteItem.kt [hunk 6]
```diff
@@ -1,8 +1,8 @@
 package com.sourcegraph.cody.vscode
 
 class InlineAutocompleteItem(
-    val insertText: String,
+    var insertText: String,
     val filterText: String,
-    val range: Range,
+    var range: Range,
     val command: Command
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/vscode/InlineAutocompleteItem.kt]: I do not like this var-val mix here 😕 maybe all these fields should be var? or it could be a data class? wdyt?

File: src/main/kotlin/com/sourcegraph/cody/editor/CodyLookupListener.kt [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            object : LookupListener {
+              override fun uiRefreshed() {
+                val selectedValue = newLookup.list.getSelectedValue()
+                if (selectedValue is LookupElement) {
+                  var lookupString = selectedValue.lookupString
+                  if (lookupString.startsWith(".")) {
+                    lookupString = lookupString.substring(1)
+                  }
+                  val offset = newEditor.caretModel.offset
+                  logger.debug("Triggering autocompletion fro lookup element: $lookupString")
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/editor/CodyLookupListener.kt]: nit: a typo - `for`

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/AutocompleteParams.kt [hunk 8]
```diff
@@ -10,5 +11,8 @@ enum class AutocompleteTriggerKind {
 data class AutocompleteParams(
     val filePath: String,
     val position: Position,
-    val triggerKind: AutocompleteTriggerKind? = AutocompleteTriggerKind.AUTOMATIC
+    val triggerKind: AutocompleteTriggerKind? = AutocompleteTriggerKind.AUTOMATIC,
+    val selectedCompletionInfo: SelectedCompletionInfo? = null
 )
+
+data class SelectedCompletionInfo(val text: String, val range: Range)
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/agent/protocol/AutocompleteParams.kt]: does it make sense to extract it to a separate file?

File: src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt [hunk 9]
```diff
@@ -27,12 +27,7 @@ import com.sourcegraph.cody.config.CodyAuthenticationManager
 import com.sourcegraph.cody.statusbar.CodyAutocompleteStatus
 import com.sourcegraph.cody.statusbar.CodyAutocompleteStatusService.Companion.notifyApplication
 import com.sourcegraph.cody.statusbar.CodyAutocompleteStatusService.Companion.resetApplication
-import com.sourcegraph.cody.vscode.CancellationToken
-import com.sourcegraph.cody.vscode.InlineAutocompleteItem
-import com.sourcegraph.cody.vscode.InlineAutocompleteList
-import com.sourcegraph.cody.vscode.InlineCompletionTriggerKind
-import com.sourcegraph.cody.vscode.IntelliJTextDocument
-import com.sourcegraph.cody.vscode.TextDocument
+import com.sourcegraph.cody.vscode.*
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt:30]: Are we ok with the wildcard imports?

Comment (Sa1to) [reply src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt:30]: I guess, linter doesn't complain about it...

**Review by mkondratek — APPROVED** [review]
